### ISSUE TITLE
Support `nbconvert --execute`

### ISSUE THREAD

**Issue by Yoshanuikabundi** [nglviewer/nglview#1063]
Hi! I'm Josh, I'm in charge of documentation at [Open Force Field](https://docs.openforcefield.org). We **love** nglview and use it all over the place - in examples, notebooks, talks, visualizing stuff during development, and in our rendered docs. Thank you for your work on it!

I'd like to see if it's possible to get a workflow involving `nbconvert --execute` to work with nglview. I've included some background info documenting where I'm up to, and then have a brief proposal that I think might fix this problem.

# Background

I'm currently setting up a cookbook to store and render all our examples so that users can navigate, access, and read them all in one place. We've used nglview in pre-rendered documentation [before](https://docs.openforcefield.org/projects/toolkit/en/stable/users/molecule_cookbook.html#from-protein-pdb-file), and I was delighted to find that it works out of the box with Sphinx as long as the widget state is saved. Last week while trying to set up automatically executing notebooks, I discovered that this does not work when the notebook is executed with `nbconvert --execute`, as a lot of the initialization occurs in JavaScript which is not run by nbconvert (see #1019, https://github.com/jupyter/nbconvert/issues/2004).

Today I dug in to your code to see if I could figure out what's going on. I couldn't get a dev environment set up - I tried to follow the instructions in CONTRIBUTING.md relating to working in JupyterLab but got a lot of warnings along the lines of "`@lumino/widgets@npm:1.37.2 doesn't provide crypto (p8f630), requested by @lumino/coreutils`", and then an error "`UserWarning: ValueError: Could not find @jupyterlab/builder`". I'll take a crack at seeing what CI does tomorrow but I've never really written anything in JS and everything I've ever tried with NPM has just worked so I don't really know what I'm doing. If someone could double check that the instructions in CONTRIBUTING.md work with a fresh clone (or whatever resets the NPM state) that would be amazing!

What I have figured out is that I can grab `nglview-js-widgets.js` from the CDN, save it to the same directory where I have the rendered notebook in HTML format, and the HTML will use the downloaded file instead of grabbing it from the CDN. With this I've experimented a bit, and it turns out I can get the NBConverted notebook to render just by adding a line to the JavaScript `handleEmbed` function:

(The CDN delivers 3.0.1, not 3.0.5, but I've ported this bit over from the current `master` too)

```javascript
            // Must call _set_representation_from_repr_dict after "fire_embed"
            // User might add Shape (buffer component) to the view and the buffer component
            // is not created yet via loadFile
            // https://github.com/nglviewer/nglview/issues/1003
            var repr_dict = that.model.get("_ngl_repr_dict");
            // Overwrite repr_dict with the default state
            repr_dict = {"0":{"0":{"params":{"aspectRatio":5,"assembly":"default","capped":true,"clipCenter":{"x":0,"y":0,"z":0},"clipNear":0,"clipRadius":0,"colorMode":"hcl","colorReverse":false,"colorScale":"RdYlBu","colorScheme":"chainname","colorValue":9474192,"defaultAssembly":"BU1","depthWrite":true,"diffuse":16777215,"diffuseInterior":false,"disablePicking":false,"flatShaded":false,"interiorColor":2236962,"interiorDarkening":0,"lazy":false,"matrix":{"elements":[1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1]},"metalness":0,"opacity":1,"quality":"high","radialSegments":20,"radiusData":{},"radiusScale":0.7,"radiusSize":1,"radiusType":"sstruc","roughness":0.4,"sele":"","side":"double","smoothSheet":false,"subdiv":6,"tension":null,"useInteriorColor":true,"visible":true,"wireframe":false},"type":"cartoon"},"1":{"params":{"aspectRatio":1,"assembly":"default","bondScale":0.4,"clipCenter":{"x":0,"y":0,"z":0},"clipNear":0,"clipRadius":0,"colorMode":"hcl","colorReverse":false,"colorScale":"RdYlBu","colorScheme":"chainname","colorValue":9474192,"cylinderOnly":false,"defaultAssembly":"BU1","depthWrite":true,"diffuse":16777215,"diffuseInterior":false,"disableImpostor":false,"disablePicking":false,"flatShaded":false,"interiorColor":2236962,"interiorDarkening":0,"lazy":false,"lineOnly":false,"linewidth":2,"matrix":{"elements":[1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1]},"metalness":0,"opacity":1,"openEnded":true,"quality":"high","radialSegments":20,"radiusData":{},"radiusScale":1,"radiusSize":0.3,"radiusType":"size","roughness":0.4,"sele":"","side":"double","sphereDetail":2,"useInteriorColor":true,"visible":true,"wireframe":false},"type":"base"},"2":{"params":{"aspectRatio":1.5,"assembly":"default","bondScale":0.3,"bondSpacing":0.75,"clipCenter":{"x":0,"y":0,"z":0},"clipNear":0,"clipRadius":0,"colorMode":"hcl","colorReverse":false,"colorScale":"","colorScheme":"element","colorValue":9474192,"cylinderOnly":false,"defaultAssembly":"BU1","depthWrite":true,"diffuse":16777215,"diffuseInterior":false,"disableImpostor":false,"disablePicking":false,"flatShaded":false,"interiorColor":2236962,"interiorDarkening":0,"lazy":false,"lineOnly":false,"linewidth":2,"matrix":{"elements":[1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1]},"metalness":0,"multipleBond":"off","opacity":1,"openEnded":true,"quality":"high","radialSegments":20,"radiusData":{},"radiusScale":2,"radiusSize":0.15,"radiusType":"size","roughness":0.4,"sele":"ligand","side":"double","sphereDetail":2,"useInteriorColor":true,"visible":true,"wireframe":false},"type":"ball+stick"}}}
            console.log("repr dict" + JSON.stringify(repr_dict));
            that._set_representation_from_repr_dict(repr_dict);
```

The camera orientation is a bit weird because it is also set by the (non-existent) state information from the notebook. Maybe it's obvious to everyone else that this works but to me it felt like a breakthrough - it demonstrates that it should be possible to get at least default representations working pretty easily on embedded widgets. So what I've surmised about how this all works is:

1. The widget is executed in Python, which involves sending a message containing the structure/trajectory from Python to JS. This message is stored in shared state between Python and JS
2. Different things happen in the notebook to `nbconvert --execute`:
   - In the notebook, JS receives the message, initializes the NGL stage, adds the object, sets up the default representations, configures a camera orientation, and stores all of that (plus anything else the user does to it) in the shared state
   - In nbconvert, no JS is executed, so nothing happens. The shared state includes only the message and the Python default values (mostly empty dicts).
3. The widget state is saved, which involves writing the shared state to the notebook file
4. `nbconvert --to html` stores the state in the HTML page
5. The converted HTML page is opened and JS runs on it. The topology/structure is successfully loaded, but when the `handleEmbed` function is called, different things happen again:
   - In the file that was executed in the notebook: representations, camera orientation, etc are present in the shared state, and are correctly loaded. The widget is reproduced.
   - In the file that was executed with nbconvert: Representations, camera orientation, etc are set to their Python defaults, NOT what JS sets them to after reading the Python defaults. The widget is displayed but has no representations.

# Proposal

If this is right, I think a simple solution might be something like:

1. Add a variable called `_ngl_is_initialized` or `_ngl_js_has_run` or something similar to the shared state, which Python sets to `False` when the widget is created.
2. if `_ngl_is_initialized` is `false`, `handleEmbed` just fires off every message in `_ngl_msg_archive` (not just the `fire_embed` ones), sets `_ngl_is_initialized` to `true`, and returns, skipping setting the model to the stored state. If `_ngl_is_initialized` is `true`, `handleEmbed` maintains its current behavior

This way, the first time the JS is run all the initialization happens, whether that's in the notebook or in the embedded context.

Does this sound right? If I can figure this out, would you accept a PR that implements this? I think it would be really great to have nglview work with a headless executor!

(Also, I'm not precious about writing this - if someones who actually understands JavaScript and NPM and everything else wants to implement this and save me the trouble of figuring all that out I'd be thrilled)

Thanks again for all your work!

**Comment by hainm** [comment]
> If someone could double check that the instructions in CONTRIBUTING.md work with a fresh clone (or whatever resets the NPM state) that would be amazing!

I think the doc is oudated.
Basically, for a fresh install:
```
cd js
npm install
cd ../
pip install . 
```

If you do not want to reinstall `nglview` after changing the code, create a symlink

```
cd ~/miniforge3/share/jupyter/labextensions # or so on
ln -sv /path/to/nglview/nglview/staticlab nglview-js-widgets
```

After that, you only need to `cd js && npm install` to update the javascript code without re-setuping the (python) nglview package.

**Comment by hainm** [comment]
> Does this sound right? If I can figure this out, would you accept a PR that implements this? I think it would be really great to have nglview work with a headless executor!

I think it's fine to have this mechanism as a hacky way to do thing.  Since it's an additive change (does not change current behaviour?), feel free to work on a PR. Cheers.

**Comment by Yoshanuikabundi** [comment]
Awesome! I'll give that a go.

**Comment by Yoshanuikabundi** [comment]
Ok I've tried the above and I get an error.

Here's exactly what I'm doing:

```shell
git clone https://github.com/nglviewer/nglview.git
cd nglview
mamba env create -n nglview-dev -f environment.yml
mamba activate nglview-dev
cd js
npm install
```

The output I get from the last command is:

```
npm WARN deprecated crypto@1.0.1: This package is no longer supported. It's now a built-in Node module. If you've depended on crypto, you should switch to the one that's built-in.
npm WARN deprecated @npmcli/move-file@1.1.2: This functionality has been moved to @npmcli/fs
npm WARN deprecated set-value@0.2.0: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.4.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.3.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.3.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.3.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.4.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.4.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.4.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.3.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.3.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.3.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.3.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated set-value@0.4.3: Critical bug fixed in v3.0.1, please upgrade to the latest version.
npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm WARN deprecated flatten@1.0.3: flatten is deprecated in favor of utility frameworks such as lodash.
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated acorn-dynamic-import@2.0.2: This is probably built in to whatever tool you're using. If you still need it... idk
npm WARN deprecated browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm WARN deprecated chokidar@2.1.8: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies
npm WARN deprecated svgo@0.7.2: This SVGO version is no longer supported. Upgrade to v2.x.x.
npm WARN deprecated popper.js@1.16.1: You can find the new Popper v2 at @popperjs/core, this package is dedicated to the legacy v1
npm WARN deprecated core-js@2.6.12: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.

> nglview-js-widgets@3.0.1 prepublish
> jlpm build:prod && webpack && jlpm run copy:labextension

Internal Error: nglview-js-widgets@workspace:.: This package doesn't seem to be present in your lockfile; run "yarn install" to update the lockfile
    at q0.getCandidates (/home/joshmitchell/conda/envs/nglview-dev/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:435:5145)
    at xf.getCandidates (/home/joshmitchell/conda/envs/nglview-dev/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:391:1264)
    at /home/joshmitchell/conda/envs/nglview-dev/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:439:7693
    at ff (/home/joshmitchell/conda/envs/nglview-dev/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:390:11070)
    at ge (/home/joshmitchell/conda/envs/nglview-dev/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:439:7673)
    at async Promise.allSettled (index 0)
    at async to (/home/joshmitchell/conda/envs/nglview-dev/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:390:10398)
    at async /home/joshmitchell/conda/envs/nglview-dev/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:439:8205
    at async ti.startProgressPromise (/home/joshmitchell/conda/envs/nglview-dev/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:390:47579)
    at async je.resolveEverything (/home/joshmitchell/conda/envs/nglview-dev/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:439:6233)
npm ERR! code 1
npm ERR! path /home/joshmitchell/Documents/nglview/js
npm ERR! command failed
npm ERR! command sh -c jlpm build:prod && webpack && jlpm run copy:labextension

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/joshmitchell/.cache/paru/clone/ytmdesktop/src/npm-cache/_logs/2023-06-14T08_16_11_658Z-debug-0.log
```

That log file is here: [2023-06-14T08_16_11_658Z-debug-0.log](https://github.com/nglviewer/nglview/files/11743398/2023-06-14T08_16_11_658Z-debug-0.log)

I get the same error if I run `yarn install` (which completes successfully, but doesn't produce the `lib` directory or any compilation artifacts) and then re-run `npm install`.

I noticed that the log file is in a wierd place but I've confirmed that the `npm` that's running comes from the `nglview-dev` conda environment.

Any ideas?

**Comment by hainm** [comment]
> Any ideas?

um, I don't really have any clue.
I've `git clean` my `js/` folder and rerun `npm install`. It's fine.
I am using the same versions as yours
```
$ npm --version
9.5.0

$ node --version
v18.15.0
```

**Comment by hainm** [comment]
```
$ jlpm --version
1.21.1
```

**Comment by Yoshanuikabundi** [comment]
It looks like I can get the extension to build by doing what CI does (`python -m pip install -U jupyterlab~=3.0 jupyter_packaging~=0.7.9 jupyterlab_widgets~=1.0.0`) - I'm just seeing if pinning those versions in environment.yml sorts things out now. `jupyter_packaging` in particular seems to be missing from my conda environment.

**Comment by hainm** [comment]
> I run `yarn install` (which completes successfully, but doesn't produce the `lib` directory or any compilation artifacts)

odd. this gave me everything.

**Comment by Yoshanuikabundi** [comment]
Looks like `jupyterlab~=3.0` is sufficient!

**Comment by hainm** [comment]
> Looks like `jupyterlab~=3.0` is sufficient!

👍 🙏

**Comment by Yoshanuikabundi** [comment]
Thanks for the help! PR is #1064

**Comment by hainm** [comment]
@Yoshanuikabundi your change is included in v3.0.6. Cheers and thanks.
